pkg/controller/node/nodecontroller.go

```diff
@@ -474,28 +529,60 @@ func (nc *NodeController) monitorNodeStatus() error {
 
 // reconcileNodeCIDRs looks at each node and assigns it a valid CIDR
 // if it doesn't currently have one.
+// Examines the availableCIDRs set first, if no more CIDR in it, generate more.
 func (nc *NodeController) reconcileNodeCIDRs(nodes *api.NodeList) {
 	glog.V(4).Infof("Reconciling cidrs for %d nodes", len(nodes.Items))
-	// TODO(roberthbailey): This seems inefficient. Why re-calculate CIDRs
-	// on each sync period?
-	availableCIDRs := generateCIDRs(nc.clusterCIDR, len(nodes.Items))
-	for _, node := range nodes.Items {
-		if node.Spec.PodCIDR != "" {
-			glog.V(4).Infof("CIDR %s is already being used by node %s", node.Spec.PodCIDR, node.Name)
-			availableCIDRs.Delete(node.Spec.PodCIDR)
-		}
-	}
-	for _, node := range nodes.Items {
-		if node.Spec.PodCIDR == "" {
-			podCIDR, found := availableCIDRs.PopAny()
-			if !found {
-				nc.recordNodeStatusChange(&node, "CIDRNotAvailable")
-				continue
+	// check if the this node controller is restarted because of crash
+	// this will only be ran once when the controller being restarted(because of crashed) or newly start
+	if nc.needSync {
+		// if it is crashed, restore the availableCIDRs by generating CIDRs, insert them into availableCIDRs set
+		// and delete assigned CIDRs from the the availableCIDRs set
+		for _, node := range nodes.Items {
+			if node.Spec.PodCIDR != "" {
+				if nc.availableCIDRs.Has(nc.translateCIDRtoIndex(node.Spec.PodCIDR)) {
+					nc.availableCIDRs.Delete(nc.translateCIDRtoIndex(node.Spec.PodCIDR))
+				} else {
+					for true {
+						result := nc.generateNewAvailabeCIDRs()
+						// CIDRs has been used up and still cannot find the node's CIDR
+						// this means the previous assigned CIDR to this node has problem, empty its CIDR so it can be reassigned
+						if !result {
+							nc.recordNodeStatusChange(&node, "CIDRNotAvailable")
```

Comment:
@gmarek , I don't quiet understand why we need to evict all Pods from this Node when the node controller restart. I did not find out what Pod information storing in node controller need to be restored when the node controller restart so that we need to evict all Pods and spawn new Pods.
@davidopp , what do you think? should we evict all Pods from all assigned CIDR Nodes in case(the node controller restart)?